### DISCUSSION TITLE
use WidgetRef from callback (using infinite list package)

### DISCUSSION THREAD

**Discussion by peterweb2005** [rrousselGit/riverpod#2227] [Q&A]
thanks

## purpose:

use the paging package
https://pub.dev/packages/infinite_scroll_pagination

but the paging event is from callback:
_pagingController.addPageRequestListener((pageKey) {

here cant use widgetRef.watch(), which can have lifecyle & family (cache) feature

so, i use a stream, then build() watch the stream provider,
so that the callback send request event, build() receive the request event, then can use widgetRef.watch(), get lifecyle & family (cache) feature

is it good? strange? any alternatives?

thanks

**Comment by peterweb2005** [comment]
## Added

after read your example, which related to an issue, about infinite list,
Your suggestions is to have a provider scope and provider override

### Implement

1. replace the stream request provider to a notifier provider
2. enclosed the page widget with a provider scope, with override with

**Comment by rrousselGit** [comment]
Wrap whatever widget you have into a consumer and you'll get a ref.

Be it adding a Consumer or converting an existing widget to a ConsumerWidget/variant

**Comment by peterweb2005** [comment]
thanks
yes, I already wrapped in stateful consumer

but maybe I no need stream provider just need state provider 

may need further investigate

**Comment by peterweb2005** [comment]
additional question, about how to retry the state provider,
for example, 2nd page has network problem, i show a button on pressed would retry,
how to implement?

read about the issues related, reach by keyword: retry

but refresh / invalidate seems not work
```
      onPressed: () {
        if (widgetRef != null && pageProvider != null) {
          widgetRef.refresh(pageProvider);
        }
      },
```

## MY BUG FIX

yes to use invalidate, i just invalidate the wrong provider

**Comment by Trung15010802** [comment]
Hi @peterweb2005 , i wanna know how do you combine riverpod + infinite package. Could you publish the source code pls?
